javaagent-tooling/src/testPatchBytecodeVersion/java/io/opentelemetry/javaagent/tooling/instrumentation/indy/PatchBytecodeVersionTest.java

```diff
@@ -0,0 +1,169 @@
+                  public void onError(
+                      String typeName,
+                      ClassLoader classLoader,
+                      JavaModule module,
+                      boolean loaded,
+                      Throwable throwable) {
+                    logger.error("Transformation error", throwable);
+                  }
+                })
+            .type(nameStartsWith(CLASSNAME_PREFIX))
+            .transform(new PatchByteCodeVersionTransformer())
+            .transform(transformerFor(isMethod().and(named("toString"))));
+
+    ByteBuddyAgent.install();
+    transformer = builder.installOn(ByteBuddyAgent.getInstrumentation());
+  }
+
+  @AfterAll
+  static void tearDown() {
+    transformer.reset(
+        ByteBuddyAgent.getInstrumentation(), AgentBuilder.RedefinitionStrategy.RETRANSFORMATION);
+
+    System.clearProperty(BYTEBUDDY_DUMP);
+  }
+
+  @ParameterizedTest
+  @MethodSource("bytecodeVersions")
+  void upgradeBytecode(ClassFileVersion version) {
+
+    String className = CLASSNAME_PREFIX + version.getMinorMajorVersion();
+
+    int startCount = PatchTestAdvice.invocationCount.get();
+    assertThat(PatchTestAdvice.invocationCount.get()).isEqualTo(startCount);
+
+    assertThat(OldBytecode.generateAndRun(className, version)).isEqualTo("toString");
+
+    assertThat(PatchTestAdvice.invocationCount.get()).isEqualTo(startCount + 1);
+
+    Path instrumentedClass;
+    Path instrumentedClassOriginal;
+
+    try (Stream<Path> files =
+        Files.find(
+            tempDir,
+            1,
+            (path, attr) -> {
+              String fileName = path.getFileName().toString();
+              return Files.isRegularFile(path)
+                  && fileName.startsWith(className)
+                  && fileName.contains(ORIGINAL_SUFFIX);
+            })) {
+
+      instrumentedClassOriginal = files.findFirst().orElseThrow(IllegalStateException::new);
+      String upgradedClassFileName =
+          instrumentedClassOriginal.getFileName().toString().replace(ORIGINAL_SUFFIX, "");
+      instrumentedClass = instrumentedClassOriginal.resolveSibling(upgradedClassFileName);
+
+    } catch (IOException e) {
+      throw new IllegalStateException(e);
+    }
+
+    assertThat(instrumentedClass).exists();
+    assertThat(instrumentedClassOriginal).exists();
+
+    assertThat(getBytecodeVersion(instrumentedClassOriginal))
+        .describedAs(
+            "expected original bytecode for class '%s' in '%s' should have been compiled for %s",
+            className, instrumentedClassOriginal.toAbsolutePath(), version)
+        .isEqualTo(version);
+
+    if (version.isLessThan(ClassFileVersion.JAVA_V7)) {
+      assertThat(getBytecodeVersion(instrumentedClass))
+          .describedAs(
+              "expected instrumented bytecode for class '%s' in '%s' should have been upgraded to Java 7",
+              className, instrumentedClass.toAbsolutePath())
+          .isEqualTo(ClassFileVersion.JAVA_V7);
+    }
```

Comment:
Nit: you could add an assertion on else that checks `assertThat(getBytecodeVersion(instrumentedClass)).isEqualTo(version)` -- this would verify that the transformer did not modify the class file version in any way.

Reply:
assertion added in e9073893b03319d205cc7caeaa973e8e3c6a57bb